lib/matplotlib/axis.py

```diff
@@ -1515,37 +1515,101 @@ def set_major_formatter(self, formatter):
+        The field used for the value must be labeled ``'x'`` and the field used
+        for the position must be labeled ``'pos'``.
+        See the  `~matplotlib.ticker.StrMethodFormatter` documentation for
+        more information.
+
+        For a function a `~matplotlib.ticker.FuncFormatter` is used.
+        The function should take in two inputs (a tick value ``x`` and a
+        position ``pos``), and return a string containing the corresponding
+        tick label.
+        See the  `~matplotlib.ticker.FuncFormatter` documentation for
+        more information.
+
         Parameters
         ----------
-        formatter : `~matplotlib.ticker.Formatter`
+        formatter : `~matplotlib.ticker.Formatter`, ``str``, or function
+        """
+        self._set_formatter(formatter, self.minor)
+
+    def _set_formatter(self, formatter, level):
         """
-        cbook._check_isinstance(mticker.Formatter, formatter=formatter)
+        Set the formatter of the major or minor ticker.
+
+        In addition to a `~matplotlib.ticker.Formatter` instance,
+        this also accepts a ``str`` or function.
+
+        For a ``str`` a `~matplotlib.ticker.StrMethodFormatter` is used.
+        The field used for the value must be labeled ``'x'`` and the field used
+        for the position must be labeled ``'pos'``.
+        See the  `~matplotlib.ticker.StrMethodFormatter` documentation for
+        more information.
+
+        For a function a `~matplotlib.ticker.FuncFormatter` is used.
+        The function should take in two inputs (a tick value ``x`` and a
+        position ``pos``), and return a string containing the corresponding
+        tick label.
+        See the  `~matplotlib.ticker.FuncFormatter` documentation for
+        more information.
+
+        Parameters
+        ----------
+        formatter : `~matplotlib.ticker.Formatter`, ``str``, or function
+        level : `self.major` or `self.minor`
+        """
+        if isinstance(formatter, str):
+            formatter = mticker.StrMethodFormatter(formatter)
+        # Don't allow any other TickHelper to avoid easy-to-make errors,
+        # like using a Locator instead of a Formatter.
+        elif (callable(formatter) and
+              not isinstance(formatter, mticker.TickHelper)):
+            formatter = mticker.FuncFormatter(formatter)
+        else:
+            cbook._check_isinstance(mticker.Formatter, formatter=formatter)
+
         if (isinstance(formatter, mticker.FixedFormatter)
                 and len(formatter.seq) > 0
-                and not isinstance(self.minor.locator, mticker.FixedLocator)):
+                and not isinstance(level.locator, mticker.FixedLocator)):
             cbook._warn_external('FixedFormatter should only be used together '
                                  'with FixedLocator')
-        self.isDefault_minfmt = False
-        self.minor.formatter = formatter
+
+        if level == self.major:
+            self.isDefault_majfmt = False
```

Comment:
really unclear what this attributes mean/do so not sure if the code is doing the right thing here

Reply:
This attribute is being used by the original, individual functions.  I just combined them.  But I added a test to make sure they are being set properly.